File: js/dailies/01_sep_18/binarySearch.js
const ArraySearchTester = require('../../helpers/testing/arrays/ArraySearchTester');

const binarySearch = (arr, x) => {
	let begIdx = 0;
	let endIdx = arr.length - 1;

	while (begIdx <= endIdx) {
		const mid = Math.floor((endIdx - begIdx) / 2) + begIdx;

		if (arr[mid] === x) return true;
		else if (arr[mid] < x) begIdx = mid + 1;
		else endIdx = mid - 1;
	}

	return false;
};

new ArraySearchTester(binarySearch);

File: python/dailies/1_nov_22/binary_search.py
from typing import List

from helpers.testing.list_search_tester import ListSearchTester


def binary_search(lst: List[int], x: int) -> bool:
    beg_idx = 0
    end_idx = len(lst) - 1

    while beg_idx <= end_idx:
        mid = ((end_idx - beg_idx) // 2) + beg_idx

        if lst[mid] == x:
            return True

        if lst[mid] < x:
            beg_idx = mid + 1
        else:
            end_idx = mid - 1

    return False


ListSearchTester(binary_search)
